python/cudf_polars/cudf_polars/experimental/explain.py

```diff
@@ -0,0 +1,171 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Explain logical and physical plans."""
+
+from __future__ import annotations
+
+import functools
+from collections.abc import MutableMapping
+from itertools import groupby
+from typing import TYPE_CHECKING
+
+from cudf_polars.dsl.ir import (
+    IR,
+    GroupBy,
+    Join,
+    Scan,
+    Sort,
+)
+from cudf_polars.dsl.translate import Translator
+from cudf_polars.experimental.base import PartitionInfo
+from cudf_polars.experimental.parallel import lower_ir_graph
+from cudf_polars.utils.config import ConfigOptions
+
+if TYPE_CHECKING:
+    import polars as pl
+
+
+def explain_query(
+    q: pl.LazyFrame, engine: pl.GPUEngine, *, physical: bool = True
+) -> str:
+    """
+    Return a formatted string representation of the IR plan.
+
+    Parameters
+    ----------
+    q : pl.LazyFrame
+        The LazyFrame to explain.
+    engine : pl.GPUEngine
+        The configured GPU engine to use.
+    physical : bool, default True
+        If True, show the physical (lowered) plan.
+        If False, show the logical (pre-lowering) plan.
+
+    Returns
+    -------
+    str
+        A string representation of the IR plan.
+    """
+    config = ConfigOptions.from_polars_engine(engine)
+    ir = Translator(q._ldf.visit(), engine).translate_ir()
+
+    if physical:
+        lowered_ir, partition_info = lower_ir_graph(ir, config)
+        return _repr_ir_tree(lowered_ir, partition_info)
+    else:
+        return _repr_ir_tree(ir)
+
+
+def _repr_ir_tree(
+    ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    header = _repr_ir(ir, partition_info, offset=offset)
```

Comment:
It seems like the logic you are using to calculate and print the partition count is the same in every `_repr_ir` dispatch. Can we remove that "physical" information from `_repr_ir` and attach it in `_repr_ir_tree`? this will make most `_repr_ir` dispatch instances pretty simple to write. The only place I see "different" logic right now is `Scan`, and that specialization doesn't seem necessary. WDYT?

Reply:
Yes that's a good point. I did some cleaning in https://github.com/rapidsai/cudf/pull/18708/commits/e542ba0b7bdbc96818e16319217893d4750c48b5